libs/constrained-generators/src/Constrained/List.hs

```diff
@@ -0,0 +1,139 @@
+
+-- | Append two type-level lists
+type family Append as as' where
+  Append '[] as' = as'
+  Append (a : as) as' = a : Append as as'
+
+-- | Map a type functor over a list
+type family MapList (f :: k -> j) as where
+  MapList f '[] = '[]
+  MapList f (a : as) = f a : MapList f as
+
+-- | A List with a hole in it, can be seen as a zipper
+-- over type-level lists.
+data ListCtx f (as :: [Type]) c where
+  (:?) :: c a -> List f as -> ListCtx f (a : as) c
+  (:!) :: f a -> ListCtx f as c -> ListCtx f (a : as) c
+
+infixr 5 :?, :!
+
+-- | A Convenient pattern for singleton contexts
+pattern NilCtx :: c a -> ListCtx f '[a] c
+pattern NilCtx c = ListCtx Nil c Nil
+
+{-# COMPLETE NilCtx #-}
+
+-- | A view of a `ListCtx` where you see the whole context at the same time.
+pattern ListCtx :: () => as'' ~ Append as (a : as') => List f as -> c a -> List f as' -> ListCtx f as'' c
+pattern ListCtx as c as' <- (toWholeCtx -> ListCtxWhole as c as')
+  where
+    ListCtx as c as' = fromWholeCtx $ ListCtxWhole as c as'
+
+{-# COMPLETE ListCtx #-}
+
+-- | Internals for the `ListCtx` pattern
+data ListCtxWhole f as c where
+  ListCtxWhole ::
+    List f as ->
+    c a ->
+    List f as' ->
+    ListCtxWhole f (Append as (a : as')) c
+
+toWholeCtx :: ListCtx f as c -> ListCtxWhole f as c
+toWholeCtx (hole :? suf) = ListCtxWhole Nil hole suf
+toWholeCtx (x :! xs)
+  | ListCtxWhole pre hole suf <- toWholeCtx xs =
+      ListCtxWhole (x :> pre) hole suf
+
+fromWholeCtx :: ListCtxWhole f as c -> ListCtx f as c
+fromWholeCtx (ListCtxWhole Nil hole suf) = hole :? suf
+fromWholeCtx (ListCtxWhole (x :> pre) hole suf) = x :! fromWholeCtx (ListCtxWhole pre hole suf)
+
+-- | A function type from `ts` to `res`:
+--    FunTy '[Int, Bool] Double = Int -> Bool -> Double
+type family FunTy ts res where
+  FunTy '[] a = a
+  FunTy (a : as) r = a -> FunTy as r
+
+-- | Materialize the shape of the type list `as`, this is very useful
+-- for avoiding having to write type classes that recurse over `as`.
+listShape :: forall as. TypeList as => List (Const ()) as
+listShape = unfoldList (\_ -> Const ())
+
+-- | Higher-order functions for working on `List`s
+class TypeList ts where
+  uncurryList :: FunTy (MapList f ts) r -> List f ts -> r
+  uncurryList_ :: (forall a. f a -> a) -> FunTy ts r -> List f ts -> r
+  curryList :: (List f ts -> r) -> FunTy (MapList f ts) r
+  curryList_ :: (forall a. a -> f a) -> (List f ts -> r) -> FunTy ts r
+  unfoldList :: (forall a as. List f as -> f a) -> List f ts
+
+instance TypeList '[] where
+  uncurryList a Nil = a
+  uncurryList_ _ a Nil = a
+  curryList f = f Nil
+  curryList_ _ f = f Nil
+  unfoldList _ = Nil
+
+instance TypeList as => TypeList (a : as) where
+  uncurryList f (a :> as) = uncurryList (f a) as
+  uncurryList_ k f (a :> as) = uncurryList_ k (f $ k a) as
+  curryList f a = curryList (\xs -> f (a :> xs))
+  curryList_ p f a = curryList_ p (\xs -> f (p a :> xs))
+  unfoldList f = let xs = unfoldList f in f xs :> xs
+
```

Comment:
There is a pattern here that people who have never seen these things often miss. It is worth pointing it out. Every instance for a class (say Foo) on TypeList as has this form:

```
instance Foo (TypeList '[]) where ...

instance Foo (TypeList as) => Foo (TypeList (a : as) where ...
```

Its like a recursive function. There are two instances, one for the Base case, and one for the inductive step.